src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,822 @@
+ *
+ * get_line(dio::filein &stream, std::string &dest)
+ *                    // Read one line and push to passed std::string
+ *
+ * get_line(dio::filein &stream, std::string &dest, char delim)
+ *                    // Read file until delim is found and push to std::string
+ *
+ * getline_nothrow(dio::filein &stream, std::string &dest)
+ *                    // Exception-free variant of get_line()
+ *
+ * getline_nothrow(dio::filein &stream, std::string &dest, char delim)
+ *                    // Exception-free variant of get_line()
+ *
+ * "cin" is a global object to interact with standard input, which provides getchar().
+ *
+ * getchar() is guaranteed to be exception-free and not to throw any exceptions anyway.
+ * Will return -1 on failure or EOF.
+ *
+ * get_line() (and all of its variants) are used to read one line from the passed filein
+ * and push it to the passed std::string. Will return 0 on EOF and -1 on failure.
+ *
+ * NOTE: get_line() erases all contents of the passed std::string when there is
+ * something in the buffer!
+ *
+ * get_line() can throw these exceptions based on the throw_on_states() function setting:
+ * On eofbit:
+ *      dio::iostream_eof("End Of File has been reached.")
+ * On buffer_failbit:
+ *      dio::iostream_error("Cannot use stream's buffer: buffer is nullptr!")
+ * On string_failbit:
+ *      dio::iostream_error("Cannot put lines into given std::string.")
+ * On posixio_errno:
+ *      dio::posixio_error(errno)
+ *
+ * Usage:
+ *
+ *      // Getting a line from stdin via getline_nothrow()
+ *      while(getline_nothrow(cin, my_string)) {
+ *          my_output.write(my_string, "\n");
+ *      }
+ *
+ *      // Getting a character from stdin via getchar()
+ *      int a_character = getchar();
+ *      if (a_character == -1) return; // Check for EOF or error
+ *      my_output.write((char)a_character, "\n");
+ */
+
+constexpr unsigned IOSTREAM_BUFSIZE = 16384; // Size of the buffers
+
+namespace dio {
+
+/*
+ * Stream state bits/vars:
+ *
+ * 1. "eofbit" indicates that the End Of File has been reached, and there is nothing
+ *    left in the buffer (filein only).
+ * 2. "buffer_failbit" indicates an attempt to use the buffer when the buffer pointer
+ *    was nullptr (e.g. failure to allocate a unique_ptr of streambuf) (fileout/filein).
+ * 3. "string_failbit" indicates a failed operation in std::string related calls
+ *    (e.g. failure in pushing a line from the buffer to the given std::string in
+ *    get_line()) (filein only).
+ * 4. "posixio_errno" indicates that a POSIX I/O function failed, and the POSIX errno
+ *    was set (fileout/filein).
+ *
+ * If all of the above bits/variables are false or 0, the stream status is "good".
+ *
+ * The current_state() function returns the stream's current status based on this table.
+ */
+enum diostates
+{
+    eofbit = 0x01,
+    buffer_failbit = 0x02,
+    string_failbit = 0x04,
+    posixio_errno = 0x08
```

Comment:
"errno" isn't a state and "posix" isn't important, could this be called "io_error"?